<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Advent-of-Code-2021" data-toc-modified-id="Advent-of-Code-2021-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Advent of Code 2021</a></span><ul class="toc-item"><li><span><a href="#Reader-Function" data-toc-modified-id="Reader-Function-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Reader Function</a></span></li><li><span><a href="#Day-1" data-toc-modified-id="Day-1-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Day 1</a></span><ul class="toc-item"><li><span><a href="#Part-1" data-toc-modified-id="Part-1-1.2.1"><span class="toc-item-num">1.2.1&nbsp;&nbsp;</span>Part 1</a></span><ul class="toc-item"><li><span><a href="#Test" data-toc-modified-id="Test-1.2.1.1"><span class="toc-item-num">1.2.1.1&nbsp;&nbsp;</span>Test</a></span></li><li><span><a href="#Real-Data" data-toc-modified-id="Real-Data-1.2.1.2"><span class="toc-item-num">1.2.1.2&nbsp;&nbsp;</span>Real Data</a></span></li></ul></li><li><span><a href="#Part-2" data-toc-modified-id="Part-2-1.2.2"><span class="toc-item-num">1.2.2&nbsp;&nbsp;</span>Part 2</a></span></li></ul></li><li><span><a href="#Day-2" data-toc-modified-id="Day-2-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Day 2</a></span><ul class="toc-item"><li><span><a href="#Part-1" data-toc-modified-id="Part-1-1.3.1"><span class="toc-item-num">1.3.1&nbsp;&nbsp;</span>Part 1</a></span></li><li><span><a href="#Part-2" data-toc-modified-id="Part-2-1.3.2"><span class="toc-item-num">1.3.2&nbsp;&nbsp;</span>Part 2</a></span></li></ul></li><li><span><a href="#Day-3" data-toc-modified-id="Day-3-1.4"><span class="toc-item-num">1.4&nbsp;&nbsp;</span>Day 3</a></span></li><li><span><a href="#Day-4" data-toc-modified-id="Day-4-1.5"><span class="toc-item-num">1.5&nbsp;&nbsp;</span>Day 4</a></span><ul class="toc-item"><li><span><a href="#Organized-Code" data-toc-modified-id="Organized-Code-1.5.1"><span class="toc-item-num">1.5.1&nbsp;&nbsp;</span>Organized Code</a></span></li></ul></li><li><span><a href="#Day-5" data-toc-modified-id="Day-5-1.6"><span class="toc-item-num">1.6&nbsp;&nbsp;</span>Day 5</a></span></li><li><span><a href="#Day-6" data-toc-modified-id="Day-6-1.7"><span class="toc-item-num">1.7&nbsp;&nbsp;</span>Day 6</a></span></li><li><span><a href="#Part-1" data-toc-modified-id="Part-1-1.8"><span class="toc-item-num">1.8&nbsp;&nbsp;</span>Part 1</a></span><ul class="toc-item"><li><span><a href="#Part-2" data-toc-modified-id="Part-2-1.8.1"><span class="toc-item-num">1.8.1&nbsp;&nbsp;</span>Part 2</a></span></li></ul></li><li><span><a href="#Day-7" data-toc-modified-id="Day-7-1.9"><span class="toc-item-num">1.9&nbsp;&nbsp;</span>Day 7</a></span><ul class="toc-item"><li><span><a href="#Part-1" data-toc-modified-id="Part-1-1.9.1"><span class="toc-item-num">1.9.1&nbsp;&nbsp;</span>Part 1</a></span><ul class="toc-item"><li><span><a href="#One-Liner" data-toc-modified-id="One-Liner-1.9.1.1"><span class="toc-item-num">1.9.1.1&nbsp;&nbsp;</span>One Liner</a></span></li></ul></li><li><span><a href="#Part-2" data-toc-modified-id="Part-2-1.9.2"><span class="toc-item-num">1.9.2&nbsp;&nbsp;</span>Part 2</a></span><ul class="toc-item"><li><span><a href="#One-Liner" data-toc-modified-id="One-Liner-1.9.2.1"><span class="toc-item-num">1.9.2.1&nbsp;&nbsp;</span>One Liner</a></span></li></ul></li></ul></li><li><span><a href="#Day-8" data-toc-modified-id="Day-8-1.10"><span class="toc-item-num">1.10&nbsp;&nbsp;</span>Day 8</a></span></li><li><span><a href="#Day-9" data-toc-modified-id="Day-9-1.11"><span class="toc-item-num">1.11&nbsp;&nbsp;</span>Day 9</a></span></li><li><span><a href="#Day-10" data-toc-modified-id="Day-10-1.12"><span class="toc-item-num">1.12&nbsp;&nbsp;</span>Day 10</a></span></li><li><span><a href="#Day-11" data-toc-modified-id="Day-11-1.13"><span class="toc-item-num">1.13&nbsp;&nbsp;</span>Day 11</a></span></li><li><span><a href="#Day-12" data-toc-modified-id="Day-12-1.14"><span class="toc-item-num">1.14&nbsp;&nbsp;</span>Day 12</a></span></li><li><span><a href="#Day-13" data-toc-modified-id="Day-13-1.15"><span class="toc-item-num">1.15&nbsp;&nbsp;</span>Day 13</a></span></li><li><span><a href="#Day-14" data-toc-modified-id="Day-14-1.16"><span class="toc-item-num">1.16&nbsp;&nbsp;</span>Day 14</a></span></li><li><span><a href="#Day-15" data-toc-modified-id="Day-15-1.17"><span class="toc-item-num">1.17&nbsp;&nbsp;</span>Day 15</a></span></li><li><span><a href="#Day-16" data-toc-modified-id="Day-16-1.18"><span class="toc-item-num">1.18&nbsp;&nbsp;</span>Day 16</a></span></li><li><span><a href="#Day-17" data-toc-modified-id="Day-17-1.19"><span class="toc-item-num">1.19&nbsp;&nbsp;</span>Day 17</a></span></li><li><span><a href="#Day-18" data-toc-modified-id="Day-18-1.20"><span class="toc-item-num">1.20&nbsp;&nbsp;</span>Day 18</a></span></li><li><span><a href="#Day-19" data-toc-modified-id="Day-19-1.21"><span class="toc-item-num">1.21&nbsp;&nbsp;</span>Day 19</a></span></li><li><span><a href="#Day-20" data-toc-modified-id="Day-20-1.22"><span class="toc-item-num">1.22&nbsp;&nbsp;</span>Day 20</a></span></li><li><span><a href="#Day-21" data-toc-modified-id="Day-21-1.23"><span class="toc-item-num">1.23&nbsp;&nbsp;</span>Day 21</a></span></li><li><span><a href="#Day-22" data-toc-modified-id="Day-22-1.24"><span class="toc-item-num">1.24&nbsp;&nbsp;</span>Day 22</a></span></li><li><span><a href="#Day-23" data-toc-modified-id="Day-23-1.25"><span class="toc-item-num">1.25&nbsp;&nbsp;</span>Day 23</a></span></li><li><span><a href="#Day-24" data-toc-modified-id="Day-24-1.26"><span class="toc-item-num">1.26&nbsp;&nbsp;</span>Day 24</a></span></li><li><span><a href="#Day-25" data-toc-modified-id="Day-25-1.27"><span class="toc-item-num">1.27&nbsp;&nbsp;</span>Day 25</a></span></li></ul></li><li><span><a href="#Preparation-Section" data-toc-modified-id="Preparation-Section-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Preparation Section</a></span><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Push-Code" data-toc-modified-id="Push-Code-2.0.1"><span class="toc-item-num">2.0.1&nbsp;&nbsp;</span>Push Code</a></span></li></ul></li><li><span><a href="#Prepare-data-files-for-all-days" data-toc-modified-id="Prepare-data-files-for-all-days-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Prepare data files for all days</a></span></li></ul></li></ul></div>

# Advent of Code 2021

## Reader Function

In [1]:

def get_data(day=1):
    """
    Returns test and real data in list format.
    Raw data should be maintained as:
        [test data]
        Split From Here
        [actual data]
    """
    file_name = f"data/day{day}.txt"
    
    with open(file_name) as fp:
        data = fp.read().strip().split("Split From Here")
        data = [d.strip().split("\n") for d in data]
        return data
get_data()

[['199', '200', '208', '210', '200', '207', '240', '269', '260', '263'],
 ['171',
  '173',
  '174',
  '163',
  '161',
  '157',
  '156',
  '154',
  '152',
  '156',
  '151',
  '153',
  '132',
  '135',
  '151',
  '143',
  '141',
  '149',
  '145',
  '147',
  '142',
  '143',
  '139',
  '141',
  '144',
  '147',
  '137',
  '144',
  '147',
  '153',
  '151',
  '153',
  '157',
  '185',
  '186',
  '185',
  '181',
  '161',
  '177',
  '179',
  '177',
  '178',
  '173',
  '175',
  '183',
  '181',
  '191',
  '189',
  '186',
  '189',
  '192',
  '191',
  '189',
  '199',
  '208',
  '218',
  '216',
  '210',
  '209',
  '208',
  '215',
  '207',
  '198',
  '202',
  '204',
  '205',
  '204',
  '203',
  '205',
  '207',
  '208',
  '209',
  '218',
  '234',
  '231',
  '245',
  '244',
  '243',
  '265',
  '264',
  '279',
  '282',
  '283',
  '274',
  '276',
  '279',
  '281',
  '279',
  '258',
  '256',
  '289',
  '297',
  '300',
  '304',
  '308',
  '298',
  '302',
  '317',
  '316',
  '315',
  '314',
  '296',
  '298',


## Day 1

[Here is the link.](https://adventofcode.com/2021/day/1)

### Part 1

In [3]:
data,data1 = get_data() 
data = list(map(int, data))
data1 = list(map(int, data1))

data

[199, 200, 208, 210, 200, 207, 240, 269, 260, 263]

#### Test

In [4]:
pd = None
res = []
for d in data:
    if pd is None:
        res.append(None)
    else:
        if pd>d:
            res.append("0")
        else:
            res.append("1")
    pd=d

In [5]:
res.count("1")

7

#### Real Data

In [6]:
pd = None
res = []
for d in data1:
    if pd is None:
        res.append(None)
    else:
        if pd>d:
            res.append("0")
        else:
            res.append("1")
    pd=d
res.count("1")

1266

### Part 2

In [7]:
w = []
wsum = []
i = 0
ps = None
for j in range(3, len(data1)+1):
    wsum.append(sum(data1[i:j]))
    i+=1

pd = None
res1 = []
for d in wsum:
    if pd is None:
        res1.append(None)
    else:
        if pd>=d:
            res1.append("0")
        else:
            res1.append("1")
    pd=d


In [8]:
res1.count("1")

1217

## Day 2

In [9]:
data,data1 = get_data(day=2)

In [10]:
data

['forward 5', 'down 5', 'forward 8', 'up 3', 'down 8', 'forward 2']

### Part 1

In [11]:
hs = 0
vs = 0

for d in data:
    k,v = d.split()
    
    if k =="forward":
        hs+=int(v)
    elif k == "down":
        vs+=int(v)
    elif k=="up":
        vs-=int(v)
    
print(hs*vs)

150


In [12]:
hs = 0
vs = 0

for d in data1:
    k,v = d.split()
    
    if k =="forward":
        hs+=int(v)
    elif k == "down":
        vs+=int(v)
    elif k=="up":
        vs-=int(v)
    
print(hs*vs)

1947824


### Part 2

In [13]:
hs = 0
vs = 0
aim = 0

for d in data1:
    k,v = d.split()
    v = int(v)
    
    if k =="forward":
        hs+=v
        vs+=aim*v
    elif k == "down":
        aim+=v
#         vs+=v
    elif k=="up":
        aim-=v
#         vs-=v
    
print(hs*vs)

1813062561


## Day 3
[Here is the link.](https://adventofcode.com/2021/day/3)
    



In [14]:
from collections import Counter
import numpy as np
data,data1 = get_data(day=3)

In [15]:
data1

['010001110001',
 '110100000001',
 '111001001011',
 '111100001000',
 '000111101001',
 '101111001001',
 '111111000110',
 '000101110010',
 '001101001010',
 '000110001101',
 '011111111110',
 '101100010001',
 '110101100001',
 '000000000110',
 '010011000011',
 '010101110010',
 '011001001001',
 '010011010111',
 '001101010110',
 '100000101000',
 '011100110010',
 '001011110001',
 '101110110011',
 '100011011101',
 '000101111011',
 '000010011000',
 '110101101111',
 '011010101100',
 '111001011010',
 '011100110011',
 '000101000111',
 '111000110101',
 '110111010000',
 '111110010001',
 '010100100110',
 '010001011010',
 '101100100001',
 '001000101010',
 '011111010010',
 '101011100110',
 '100111010110',
 '101000110100',
 '011010110001',
 '000010001000',
 '100001110111',
 '101001011100',
 '100100001011',
 '111010101110',
 '001101101001',
 '000111111100',
 '011011111000',
 '111111111011',
 '101100011010',
 '011101001001',
 '111001001000',
 '010101010001',
 '100000011011',
 '100110100011',
 '010101110111

In [16]:
def part1(inp):
    cs = len(inp[0])
    dt = [int(d) for dt in inp for d in dt]
    dt = np.array(dt).reshape(-1, cs)

    print(dt[0])

    # sorted(x.items(), key=lambda item: item[1])
    counts = [sorted(dict(Counter(dt[:, i])).items(), key=lambda item: item[1]) for i in range(len(dt[0]))]
    counts = np.array(counts).reshape(-1,2)

#     print(counts)

    minidx = np.arange(0, len(counts), 2)
    maxidx = np.arange(1, len(counts), 2)

    minv = int("".join(list(map(str, counts[minidx, 0]))), 2)
    maxv = int("".join(list(map(str, counts[maxidx, 0]))), 2)
    print(minv, maxv)
    print(minv*maxv)
part1(data)

[0 0 1 0 0]
9 22
198


In [17]:
part1(data1)

[0 1 0 0 0 1 1 1 0 0 0 1]
1749 2346
4103154


In [18]:
def o2(dt):
    ndt = dt.copy()
    #print(len(ndt[0]))
    
    curr_c = 0
    while curr_c<len(dt[0]):
        print(f"Current Col: {curr_c}")
        counts = [sorted(dict(Counter(ndt[:, curr_c])).items(), key=lambda item: item[1])]
        counts = np.array(counts).reshape(-1,2)
        if len(counts)>1:
            if counts[0, 1]==counts[1, 1]:
                ndt = ndt[ndt[:,curr_c]==1]
            else:
                ndt = ndt[ndt[:,curr_c]==counts[1][0]]
        else:
            ndt = ndt[ndt[:,curr_c]==counts[1][0]]
        print(f"Current Col: {curr_c} Rows: {len(ndt)}")
#         print(counts)
#         print(ndt[ndt[:,curr_c]==counts[1][0]])
            
        
        
        curr_c+=1
    res = int("".join(list(map(str, ndt[0]))), 2)
    print(res)
    return res

def co2(dt):
    ndt = dt.copy()
    #print(len(ndt[0]))
    
    curr_c = 0
    while curr_c<len(dt[0]):
        
        counts = [sorted(dict(Counter(ndt[:, curr_c])).items(), key=lambda item: item[1])]
        counts = np.array(counts).reshape(-1,2)
        if len(counts)>1:
            if counts[0, 1]== counts[1, 1]:
                ndt = ndt[ndt[:,curr_c]==0]
            else:
                ndt = ndt[ndt[:,curr_c]==counts[0][0]]
        else:
            ndt = ndt[ndt[:,curr_c]==counts[0][0]]
            
#         print(ndt)
        print(f"Current Col: {curr_c} Rows: {len(ndt)}")
        curr_c+=1
    return int("".join(list(map(str, ndt[0]))), 2)

def part2(inp):
    cs = len(inp[0])
    dt = [int(d) for dt in inp for d in dt]
    dt = np.array(dt).reshape(-1, cs)
    
    print("O2")
    o2v = o2(dt)
    co2v = co2(dt)
    print(o2v, co2v)
    print(o2v*co2v)
part2(data)
    

O2
Current Col: 0
Current Col: 0 Rows: 7
Current Col: 1
Current Col: 1 Rows: 4
Current Col: 2
Current Col: 2 Rows: 3
Current Col: 3
Current Col: 3 Rows: 2
Current Col: 4
Current Col: 4 Rows: 1
23
Current Col: 0 Rows: 5
Current Col: 1 Rows: 2
Current Col: 2 Rows: 1
Current Col: 3 Rows: 1
Current Col: 4 Rows: 1
23 10
230


In [19]:
part2(data1)

O2
Current Col: 0
Current Col: 0 Rows: 511
Current Col: 1
Current Col: 1 Rows: 263
Current Col: 2
Current Col: 2 Rows: 134
Current Col: 3
Current Col: 3 Rows: 71
Current Col: 4
Current Col: 4 Rows: 36
Current Col: 5
Current Col: 5 Rows: 20
Current Col: 6
Current Col: 6 Rows: 12
Current Col: 7
Current Col: 7 Rows: 8
Current Col: 8
Current Col: 8 Rows: 5
Current Col: 9
Current Col: 9 Rows: 3
Current Col: 10
Current Col: 10 Rows: 2
Current Col: 11
Current Col: 11 Rows: 1
3399
Current Col: 0 Rows: 489
Current Col: 1 Rows: 235
Current Col: 2 Rows: 109
Current Col: 3 Rows: 54
Current Col: 4 Rows: 24
Current Col: 5 Rows: 11
Current Col: 6 Rows: 4
Current Col: 7 Rows: 2
Current Col: 8 Rows: 1
Current Col: 9 Rows: 1
Current Col: 10 Rows: 1
Current Col: 11 Rows: 1
3399 1249
4245351


## Day 4

[Here is the link.](https://adventofcode.com/2021/day/4)


In [20]:
import numpy as np
data,data1 = get_data(day=4)

In [21]:
data

['7,4,9,5,11,17,23,2,0,14,21,24,10,16,13,6,15,25,12,22,18,20,8,19,3,26,1',
 '',
 '22 13 17 11  0',
 ' 8  2 23  4 24',
 '21  9 14 16  7',
 ' 6 10  3 18  5',
 ' 1 12 20 15 19',
 '',
 ' 3 15  0  2 22',
 ' 9 18 13 17  5',
 '19  8  7 25 23',
 '20 11 10 24  4',
 '14 21 16 12  6',
 '',
 '14 21 17 24  4',
 '10 16 15  9 19',
 '18  8 23 26 20',
 '22 11 13  6  5',
 ' 2  0 12  3  7']

In [22]:
block = []
num = [int(i) for i in data[0].split(",")]
row = []
tdata=[]
blocks = 0

for d in data[2:]:
#     print(d)
    if d == "":
        tdata.append(block)
        block=[]
        blocks+=1
        
    else:
#         print([i for i in d.strip().split(" ") if i!=""])
        block.append([int(i) for i in d.strip().split(" ") if i!=""])
tdata.append(block)
block=[]
blocks+=1
print(blocks)

3


In [23]:
tdata = np.array(tdata).reshape(blocks,-1, 5)
tdata

array([[[22, 13, 17, 11,  0],
        [ 8,  2, 23,  4, 24],
        [21,  9, 14, 16,  7],
        [ 6, 10,  3, 18,  5],
        [ 1, 12, 20, 15, 19]],

       [[ 3, 15,  0,  2, 22],
        [ 9, 18, 13, 17,  5],
        [19,  8,  7, 25, 23],
        [20, 11, 10, 24,  4],
        [14, 21, 16, 12,  6]],

       [[14, 21, 17, 24,  4],
        [10, 16, 15,  9, 19],
        [18,  8, 23, 26, 20],
        [22, 11, 13,  6,  5],
        [ 2,  0, 12,  3,  7]]])

In [24]:
results = np.zeros_like(tdata).astype(np.bool)
matched = False
for n in num:
#     print(n)
    for i,block in enumerate(tdata):
        results[i] += block==n
#         print(results[i])
        
        # search across row
        if (results[i]==[ True,  True,  True,  True,  True]).all(axis=1).any():
            print(f"Row Matched Block:{i}")
            matched=True
            break
        
        # search across cols
        if (results[i].T==[ True,  True,  True,  True,  True]).all(axis=1).any():
            print(f"Col Matched Block: {i}")
            matched=True
            break
    if matched:
#         print(i)
#         print(tdata[i])
        s = (tdata[i]*~results[i]).sum()
        print(s)
        print(n)
        print(n*s)
        break
        
        

Row Matched Block:2
188
24
4512


In [25]:
block = []
num = [int(i) for i in data1[0].split(",")]
row = []
tdata=[]
blocks = 0

for d in data1[2:]:
#     print(d)
    if d == "":
        tdata.append(block)
        block=[]
        blocks+=1
        
    else:
#         print([i for i in d.strip().split(" ") if i!=""])
        block.append([int(i) for i in d.strip().split(" ") if i!=""])
tdata.append(block)
block=[]
blocks+=1
print(blocks)

tdata = np.array(tdata).reshape(blocks,-1, 5)
tdata

results = np.zeros_like(tdata).astype(np.bool)
matched = False
mblocks=[]
all_blocks = list(range(0, blocks))
# print(all_blocks)
for n in num:
#     print(n)
    for i,block in enumerate(tdata):
        results[i] += block==n
#         print(results[i])
        
        # search across row
        if (results[i]==[ True,  True,  True,  True,  True]).all(axis=1).any():
            print(f"Row Matched Block:{i}")
            if i not in mblocks:
                mblocks.append(i)
            if len(mblocks) == len(all_blocks):
                matched=True
            
#             break
        
        # search across cols
        if (results[i].T==[ True,  True,  True,  True,  True]).all(axis=1).any():
            print(f"Col Matched Block: {i}")
            if i not in mblocks:
                mblocks.append(i)
            if len(mblocks) == len(all_blocks):
                matched=True
            
#             break
    if matched:
        i = mblocks[i]
        
        print(tdata[i])
        s = (tdata[i]*~results[i]).sum()
        print(s)
        print(n)
        print(n*s)
        break
        
        

100
Row Matched Block:73
Row Matched Block:73
Row Matched Block:73
Row Matched Block:73
Row Matched Block:73
Col Matched Block: 57
Row Matched Block:73
Col Matched Block: 57
Row Matched Block:73
Row Matched Block:79
Col Matched Block: 33
Col Matched Block: 57
Row Matched Block:73
Row Matched Block:79
Col Matched Block: 33
Col Matched Block: 57
Row Matched Block:73
Row Matched Block:79
Col Matched Block: 33
Col Matched Block: 57
Row Matched Block:73
Row Matched Block:79
Col Matched Block: 29
Col Matched Block: 33
Col Matched Block: 57
Row Matched Block:73
Row Matched Block:79
Col Matched Block: 29
Col Matched Block: 33
Col Matched Block: 57
Row Matched Block:73
Row Matched Block:79
Row Matched Block:94
Col Matched Block: 29
Col Matched Block: 33
Col Matched Block: 57
Row Matched Block:73
Row Matched Block:79
Row Matched Block:94
Row Matched Block:23
Col Matched Block: 29
Col Matched Block: 33
Col Matched Block: 57
Col Matched Block: 60
Col Matched Block: 70
Row Matched Block:73
Row Matc

Row Matched Block:57
Col Matched Block: 57
Row Matched Block:59
Col Matched Block: 59
Col Matched Block: 60
Row Matched Block:61
Col Matched Block: 61
Col Matched Block: 64
Row Matched Block:66
Row Matched Block:70
Col Matched Block: 70
Row Matched Block:72
Row Matched Block:73
Col Matched Block: 77
Row Matched Block:79
Col Matched Block: 79
Row Matched Block:80
Row Matched Block:81
Col Matched Block: 81
Row Matched Block:84
Row Matched Block:90
Row Matched Block:93
Row Matched Block:94
Row Matched Block:0
Col Matched Block: 2
Row Matched Block:3
Col Matched Block: 3
Col Matched Block: 6
Row Matched Block:8
Row Matched Block:9
Col Matched Block: 9
Row Matched Block:12
Col Matched Block: 13
Col Matched Block: 15
Row Matched Block:16
Col Matched Block: 16
Row Matched Block:17
Col Matched Block: 17
Row Matched Block:23
Col Matched Block: 23
Col Matched Block: 26
Row Matched Block:28
Col Matched Block: 28
Col Matched Block: 29
Row Matched Block:30
Row Matched Block:32
Row Matched Block:33


Row Matched Block:28
Col Matched Block: 28
Col Matched Block: 29
Row Matched Block:30
Row Matched Block:32
Row Matched Block:33
Col Matched Block: 33
Col Matched Block: 35
Row Matched Block:36
Col Matched Block: 38
Col Matched Block: 39
Col Matched Block: 40
Row Matched Block:41
Row Matched Block:43
Col Matched Block: 43
Row Matched Block:44
Col Matched Block: 44
Row Matched Block:46
Row Matched Block:48
Col Matched Block: 48
Row Matched Block:49
Col Matched Block: 50
Row Matched Block:51
Row Matched Block:53
Col Matched Block: 56
Row Matched Block:57
Col Matched Block: 57
Row Matched Block:59
Col Matched Block: 59
Row Matched Block:60
Col Matched Block: 60
Row Matched Block:61
Col Matched Block: 61
Col Matched Block: 64
Row Matched Block:66
Row Matched Block:67
Row Matched Block:70
Col Matched Block: 70
Col Matched Block: 71
Row Matched Block:72
Row Matched Block:73
Col Matched Block: 73
Row Matched Block:74
Col Matched Block: 74
Col Matched Block: 77
Row Matched Block:78
Row Matched 

Col Matched Block: 79
Row Matched Block:80
Row Matched Block:81
Col Matched Block: 81
Row Matched Block:82
Row Matched Block:84
Row Matched Block:87
Col Matched Block: 87
Row Matched Block:88
Row Matched Block:89
Row Matched Block:90
Col Matched Block: 90
Row Matched Block:91
Col Matched Block: 91
Row Matched Block:93
Col Matched Block: 93
Row Matched Block:94
Col Matched Block: 94
Row Matched Block:96
Col Matched Block: 97
Col Matched Block: 98
Col Matched Block: 99
Row Matched Block:0
Row Matched Block:1
Row Matched Block:2
Col Matched Block: 2
Row Matched Block:3
Col Matched Block: 3
Col Matched Block: 6
Col Matched Block: 7
Row Matched Block:8
Row Matched Block:9
Col Matched Block: 9
Row Matched Block:12
Col Matched Block: 12
Row Matched Block:13
Col Matched Block: 13
Row Matched Block:14
Col Matched Block: 14
Row Matched Block:15
Col Matched Block: 15
Row Matched Block:16
Col Matched Block: 16
Row Matched Block:17
Col Matched Block: 17
Col Matched Block: 19
Row Matched Block:20
Co

Row Matched Block:53
Row Matched Block:54
Row Matched Block:55
Col Matched Block: 55
Row Matched Block:56
Col Matched Block: 56
Row Matched Block:57
Col Matched Block: 57
Row Matched Block:58
Col Matched Block: 58
Row Matched Block:59
Col Matched Block: 59
Row Matched Block:60
Col Matched Block: 60
Row Matched Block:61
Col Matched Block: 61
Row Matched Block:62
Col Matched Block: 64
Row Matched Block:66
Row Matched Block:67
Col Matched Block: 67
Row Matched Block:68
Col Matched Block: 68
Row Matched Block:69
Col Matched Block: 69
Row Matched Block:70
Col Matched Block: 70
Row Matched Block:71
Col Matched Block: 71
Row Matched Block:72
Row Matched Block:73
Col Matched Block: 73
Row Matched Block:74
Col Matched Block: 74
Row Matched Block:75
Row Matched Block:76
Col Matched Block: 76
Row Matched Block:77
Col Matched Block: 77
Row Matched Block:78
Col Matched Block: 78
Row Matched Block:79
Col Matched Block: 79
Row Matched Block:80
Col Matched Block: 80
Row Matched Block:81
Col Matched Bl

Row Matched Block:79
Col Matched Block: 79
Row Matched Block:80
Col Matched Block: 80
Row Matched Block:81
Col Matched Block: 81
Row Matched Block:82
Col Matched Block: 82
Row Matched Block:83
Row Matched Block:84
Row Matched Block:86
Col Matched Block: 86
Row Matched Block:87
Col Matched Block: 87
Row Matched Block:88
Col Matched Block: 88
Row Matched Block:89
Col Matched Block: 89
Row Matched Block:90
Col Matched Block: 90
Row Matched Block:91
Col Matched Block: 91
Row Matched Block:93
Col Matched Block: 93
Row Matched Block:94
Col Matched Block: 94
Row Matched Block:96
Row Matched Block:97
Col Matched Block: 97
Row Matched Block:98
Col Matched Block: 98
Row Matched Block:99
Col Matched Block: 99
Row Matched Block:0
Col Matched Block: 0
Row Matched Block:1
Col Matched Block: 1
Row Matched Block:2
Col Matched Block: 2
Row Matched Block:3
Col Matched Block: 3
Row Matched Block:4
Col Matched Block: 4
Row Matched Block:5
Col Matched Block: 5
Row Matched Block:6
Col Matched Block: 6
Row M

Row Matched Block:8
Col Matched Block: 8
Row Matched Block:9
Col Matched Block: 9
Row Matched Block:10
Col Matched Block: 10
Col Matched Block: 11
Row Matched Block:12
Col Matched Block: 12
Row Matched Block:13
Col Matched Block: 13
Row Matched Block:14
Col Matched Block: 14
Row Matched Block:15
Col Matched Block: 15
Row Matched Block:16
Col Matched Block: 16
Row Matched Block:17
Col Matched Block: 17
Row Matched Block:18
Row Matched Block:19
Col Matched Block: 19
Row Matched Block:20
Col Matched Block: 20
Col Matched Block: 21
Col Matched Block: 22
Row Matched Block:23
Col Matched Block: 23
Row Matched Block:24
Col Matched Block: 24
Row Matched Block:25
Row Matched Block:26
Col Matched Block: 26
Row Matched Block:27
Col Matched Block: 27
Row Matched Block:28
Col Matched Block: 28
Row Matched Block:29
Col Matched Block: 29
Row Matched Block:30
Col Matched Block: 30
Row Matched Block:31
Col Matched Block: 31
Row Matched Block:32
Col Matched Block: 32
Row Matched Block:33
Col Matched Blo

Row Matched Block:16
Col Matched Block: 16
Row Matched Block:17
Col Matched Block: 17
Row Matched Block:18
Row Matched Block:19
Col Matched Block: 19
Row Matched Block:20
Col Matched Block: 20
Col Matched Block: 21
Row Matched Block:22
Col Matched Block: 22
Row Matched Block:23
Col Matched Block: 23
Row Matched Block:24
Col Matched Block: 24
Row Matched Block:25
Row Matched Block:26
Col Matched Block: 26
Row Matched Block:27
Col Matched Block: 27
Row Matched Block:28
Col Matched Block: 28
Row Matched Block:29
Col Matched Block: 29
Row Matched Block:30
Col Matched Block: 30
Row Matched Block:31
Col Matched Block: 31
Row Matched Block:32
Col Matched Block: 32
Row Matched Block:33
Col Matched Block: 33
Col Matched Block: 34
Row Matched Block:35
Col Matched Block: 35
Row Matched Block:36
Col Matched Block: 36
Row Matched Block:37
Col Matched Block: 37
Row Matched Block:38
Col Matched Block: 38
Row Matched Block:39
Col Matched Block: 39
Row Matched Block:40
Col Matched Block: 40
Row Matched

### Organized Code

In [26]:
def get_blocks(dt):
    block = []
    num = [int(i) for i in dt[0].split(",")]
    row = []
    tdata=[]
    blocks = 0
    for d in dt[2:]:
        if d == "":
            tdata.append(block)
            block=[]
            blocks+=1

        else:
            block.append([int(i) for i in d.strip().split(" ") if i!=""])
    tdata.append(block)
    block=[]
    blocks+=1
    tdata = np.array(tdata).reshape(blocks,-1, 5)
    return tdata, num

def get_first_matched(tdata, num):
    results = np.zeros_like(tdata).astype(np.bool)
    matched = False
    
    for n in num:
        for i,block in enumerate(tdata):
            results[i] += block==n
            # search across row
            if (results[i]==[ True,  True,  True,  True,  True]).all(axis=1).any():
                print(f"Row Matched Block:{i}")
                matched=True
                break

            # search across cols
            if (results[i].T==[ True,  True,  True,  True,  True]).all(axis=1).any():
                print(f"Col Matched Block: {i}")
                matched=True
                break
        if matched:
            print(f"\nResult Block: {tdata[i]}")
            s = (tdata[i]*~results[i]).sum()
            print(f"Sum: {s}")
            print(f"Last number: {n}")
            print(f"Answer: {n*s}\n")
            break

def get_last_matched(tdata, num):
    results = np.zeros_like(tdata).astype(np.bool)
    matched = False
    mblocks=[]
    all_blocks = list(range(0, len(results)))
    
    for n in num:
        for i,block in enumerate(tdata):
            results[i] += block==n
            # search across row
            if (results[i]==[ True,  True,  True,  True,  True]).all(axis=1).any():
                print(f"Row Matched Block:{i}")
                if i not in mblocks:
                    mblocks.append(i)
                if len(mblocks) == len(all_blocks):
                    matched=True

            # search across cols
            if (results[i].T==[ True,  True,  True,  True,  True]).all(axis=1).any():
                print(f"Col Matched Block: {i}")
                if i not in mblocks:
                    mblocks.append(i)
                if len(mblocks) == len(all_blocks):
                    matched=True

        if matched:
            i = mblocks[i]
            
            print(i)
            print(f"\nResult Block: {tdata[i]}")
            s = (tdata[i]*~results[i]).sum()
            print(f"Sum: {s}")
            print(f"Last number: {n}")
            print(f"Answer: {n*s}")
            break

d1,n1 = get_blocks(data1)
get_first_matched(tdata=d1, num=n1)
get_last_matched(tdata=d1, num=n1)
# d1, n1

Row Matched Block:73

Result Block: [[26 89 27 47 91]
 [15  9 18 62 28]
 [31 96 42 81 86]
 [11 52 20 93 38]
 [83 64 39  1 60]]
Sum: 782
Last number: 42
Answer: 32844

Row Matched Block:73
Row Matched Block:73
Row Matched Block:73
Row Matched Block:73
Row Matched Block:73
Col Matched Block: 57
Row Matched Block:73
Col Matched Block: 57
Row Matched Block:73
Row Matched Block:79
Col Matched Block: 33
Col Matched Block: 57
Row Matched Block:73
Row Matched Block:79
Col Matched Block: 33
Col Matched Block: 57
Row Matched Block:73
Row Matched Block:79
Col Matched Block: 33
Col Matched Block: 57
Row Matched Block:73
Row Matched Block:79
Col Matched Block: 29
Col Matched Block: 33
Col Matched Block: 57
Row Matched Block:73
Row Matched Block:79
Col Matched Block: 29
Col Matched Block: 33
Col Matched Block: 57
Row Matched Block:73
Row Matched Block:79
Row Matched Block:94
Col Matched Block: 29
Col Matched Block: 33
Col Matched Block: 57
Row Matched Block:73
Row Matched Block:79
Row Matched Block:

Row Matched Block:0
Col Matched Block: 2
Row Matched Block:3
Col Matched Block: 3
Col Matched Block: 6
Row Matched Block:8
Row Matched Block:9
Col Matched Block: 9
Row Matched Block:12
Col Matched Block: 12
Col Matched Block: 13
Col Matched Block: 15
Row Matched Block:16
Col Matched Block: 16
Row Matched Block:17
Col Matched Block: 17
Row Matched Block:20
Row Matched Block:23
Col Matched Block: 23
Row Matched Block:24
Col Matched Block: 24
Row Matched Block:26
Col Matched Block: 26
Row Matched Block:27
Col Matched Block: 27
Row Matched Block:28
Col Matched Block: 28
Col Matched Block: 29
Row Matched Block:30
Row Matched Block:32
Row Matched Block:33
Col Matched Block: 33
Col Matched Block: 35
Row Matched Block:36
Col Matched Block: 40
Row Matched Block:41
Row Matched Block:43
Col Matched Block: 43
Row Matched Block:44
Col Matched Block: 44
Row Matched Block:46
Col Matched Block: 48
Row Matched Block:49
Col Matched Block: 50
Row Matched Block:51
Row Matched Block:53
Col Matched Block: 5

Row Matched Block:27
Col Matched Block: 27
Row Matched Block:28
Col Matched Block: 28
Row Matched Block:29
Col Matched Block: 29
Row Matched Block:30
Row Matched Block:32
Col Matched Block: 32
Row Matched Block:33
Col Matched Block: 33
Col Matched Block: 35
Row Matched Block:36
Col Matched Block: 36
Col Matched Block: 37
Col Matched Block: 38
Col Matched Block: 39
Col Matched Block: 40
Row Matched Block:41
Row Matched Block:42
Row Matched Block:43
Col Matched Block: 43
Row Matched Block:44
Col Matched Block: 44
Row Matched Block:46
Col Matched Block: 46
Row Matched Block:48
Col Matched Block: 48
Row Matched Block:49
Col Matched Block: 49
Col Matched Block: 50
Row Matched Block:51
Row Matched Block:53
Row Matched Block:54
Row Matched Block:55
Col Matched Block: 55
Row Matched Block:56
Col Matched Block: 56
Row Matched Block:57
Col Matched Block: 57
Row Matched Block:59
Col Matched Block: 59
Row Matched Block:60
Col Matched Block: 60
Row Matched Block:61
Col Matched Block: 61
Col Matched

Col Matched Block: 33
Col Matched Block: 35
Row Matched Block:36
Col Matched Block: 36
Col Matched Block: 37
Col Matched Block: 38
Col Matched Block: 39
Row Matched Block:40
Col Matched Block: 40
Row Matched Block:41
Row Matched Block:42
Col Matched Block: 42
Row Matched Block:43
Col Matched Block: 43
Row Matched Block:44
Col Matched Block: 44
Row Matched Block:46
Col Matched Block: 46
Row Matched Block:48
Col Matched Block: 48
Row Matched Block:49
Col Matched Block: 49
Row Matched Block:50
Col Matched Block: 50
Row Matched Block:51
Row Matched Block:52
Col Matched Block: 52
Row Matched Block:53
Row Matched Block:54
Row Matched Block:55
Col Matched Block: 55
Row Matched Block:56
Col Matched Block: 56
Row Matched Block:57
Col Matched Block: 57
Row Matched Block:58
Col Matched Block: 58
Row Matched Block:59
Col Matched Block: 59
Row Matched Block:60
Col Matched Block: 60
Row Matched Block:61
Col Matched Block: 61
Row Matched Block:62
Col Matched Block: 64
Row Matched Block:66
Row Matched

Row Matched Block:16
Col Matched Block: 16
Row Matched Block:17
Col Matched Block: 17
Row Matched Block:18
Row Matched Block:19
Col Matched Block: 19
Row Matched Block:20
Col Matched Block: 20
Col Matched Block: 21
Col Matched Block: 22
Row Matched Block:23
Col Matched Block: 23
Row Matched Block:24
Col Matched Block: 24
Row Matched Block:25
Row Matched Block:26
Col Matched Block: 26
Row Matched Block:27
Col Matched Block: 27
Row Matched Block:28
Col Matched Block: 28
Row Matched Block:29
Col Matched Block: 29
Row Matched Block:30
Col Matched Block: 30
Row Matched Block:31
Col Matched Block: 31
Row Matched Block:32
Col Matched Block: 32
Row Matched Block:33
Col Matched Block: 33
Col Matched Block: 34
Row Matched Block:35
Col Matched Block: 35
Row Matched Block:36
Col Matched Block: 36
Row Matched Block:37
Col Matched Block: 37
Col Matched Block: 38
Col Matched Block: 39
Row Matched Block:40
Col Matched Block: 40
Row Matched Block:41
Col Matched Block: 41
Row Matched Block:42
Col Matche

Col Matched Block: 93
Row Matched Block:94
Col Matched Block: 94
Row Matched Block:95
Row Matched Block:96
Col Matched Block: 96
Row Matched Block:97
Col Matched Block: 97
Row Matched Block:98
Col Matched Block: 98
Row Matched Block:99
Col Matched Block: 99
Row Matched Block:0
Col Matched Block: 0
Row Matched Block:1
Col Matched Block: 1
Row Matched Block:2
Col Matched Block: 2
Row Matched Block:3
Col Matched Block: 3
Row Matched Block:4
Col Matched Block: 4
Row Matched Block:5
Col Matched Block: 5
Row Matched Block:6
Col Matched Block: 6
Row Matched Block:7
Col Matched Block: 7
Row Matched Block:8
Col Matched Block: 8
Row Matched Block:9
Col Matched Block: 9
Row Matched Block:10
Col Matched Block: 10
Row Matched Block:11
Col Matched Block: 11
Row Matched Block:12
Col Matched Block: 12
Row Matched Block:13
Col Matched Block: 13
Row Matched Block:14
Col Matched Block: 14
Row Matched Block:15
Col Matched Block: 15
Row Matched Block:16
Col Matched Block: 16
Row Matched Block:17
Col Matche

Col Matched Block: 50
Row Matched Block:51
Col Matched Block: 51
Row Matched Block:52
Col Matched Block: 52
Row Matched Block:53
Col Matched Block: 53
Row Matched Block:54
Col Matched Block: 54
Row Matched Block:55
Col Matched Block: 55
Row Matched Block:56
Col Matched Block: 56
Row Matched Block:57
Col Matched Block: 57
Row Matched Block:58
Col Matched Block: 58
Row Matched Block:59
Col Matched Block: 59
Row Matched Block:60
Col Matched Block: 60
Row Matched Block:61
Col Matched Block: 61
Row Matched Block:62
Col Matched Block: 62
Row Matched Block:63
Col Matched Block: 63
Row Matched Block:64
Col Matched Block: 64
Row Matched Block:66
Col Matched Block: 66
Row Matched Block:67
Col Matched Block: 67
Row Matched Block:68
Col Matched Block: 68
Row Matched Block:69
Col Matched Block: 69
Row Matched Block:70
Col Matched Block: 70
Row Matched Block:71
Col Matched Block: 71
Row Matched Block:72
Row Matched Block:73
Col Matched Block: 73
Row Matched Block:74
Col Matched Block: 74
Row Matched

## Day 5

[Here is the link.](https://adventofcode.com/2021/day/5)




In [27]:
import numpy as np

data,data1 = get_data(day=5)

In [28]:
data

['0,9 -> 5,9',
 '8,0 -> 0,8',
 '9,4 -> 3,4',
 '2,2 -> 2,1',
 '7,0 -> 7,4',
 '6,4 -> 2,0',
 '0,9 -> 2,9',
 '3,4 -> 1,4',
 '0,0 -> 8,8',
 '5,5 -> 8,2']

In [29]:
# get (x1, y1), (x2, y2)
coordinates = []
for d in data1:
    x1, y1, x2, y2 = d.replace(" -> ", ",").split(",")
    coordinates.append((x1, y1, x2, y2))
coordinates = np.array(coordinates).astype(np.uint8)


In [30]:
coordinates

array([[ 87,  19,  87,  32],
       [179, 136, 179, 233],
       [  0, 202,   0, 187],
       ...,
       [ 83,  97, 239,  97],
       [ 46,  73,  46, 105],
       [212, 183, 173, 183]], dtype=uint8)

In [31]:
coordinates = []
for d in data:
    x1, y1, x2, y2 = list(map(int, d.replace(" -> ", ",").split(",")))
    coordinates.append((x1, y1, x2, y2))
    
coordinates = np.array(coordinates)

mxx,mxy = coordinates[[0, 2]].max(), coordinates[[1, 3]].max()

board = np.zeros((mxx*2, mxy*2))

# check only horizontal or vertical line
m1 = coordinates[:, 0]==coordinates[:, 2]
m2 = coordinates[:, 1]==coordinates[:, 3]
m = m1 | m2

masked = coordinates[m]

for co in masked:    
    for x in range(min(co[0], co[2]), max(co[0], co[2])+1):
        for y in range(min(co[1], co[3]), max(co[1], co[3])+1):
            board[x, y] += 1
print((board.flatten()>1).sum())

# diagonal line
m1 = coordinates[:, 0]!=coordinates[:, 2]
m2 = coordinates[:, 1]!=coordinates[:, 3]
m=m1*m2
masked = coordinates[m]

for co in masked:
    # add or sub to x1?
    dx = int(co[2]>co[0]) or -1
    dy = int(co[3]>co[1]) or -1
    
    for dp in range(abs(co[2]-co[0])+1):
        x = co[0]+dx*dp
        y = co[1]+dy*dp
        board[x,y]+=1
        
print((board.flatten()>1).sum())

5
12


## Day 6

[Here is the link.](https://adventofcode.com/2021/day/6)




In [32]:
import numpy as np

data,data1 = get_data(day=6)

In [33]:
data

['3,4,3,1,2']

In [34]:
data = [int(d) for d in data[0].split(",")]
data

[3, 4, 3, 1, 2]

In [35]:
data1 = [int(d) for d in data1[0].split(",")]
# data1

## Part 1

In [36]:

total_days = 81
curr_data = data1.copy()
for day in range(1, total_days):    
    temp_data = []
    new_fish = []
    for d in curr_data:
        if d == 0:
            new_fish.append(8)
            d=6
        else:
            d-=1
        temp_data.append(d)
    temp_data.extend(new_fish)
    curr_data = temp_data
print(f"Total Fish: {len(curr_data)}\n")
        
        

Total Fish: 388419



In [37]:
import numpy as np

total_days = 81
curr_data = np.array(data1.copy())
for day in range(1, total_days):

    curr_data-=1
    s = sum(curr_data<0)
    curr_data[curr_data<0]=6
    curr_data = curr_data.tolist()
    curr_data.extend(s*[8])
    curr_data = np.array(curr_data)        
print(f"Total Fish: {len(curr_data)}\n")

Total Fish: 388419



### Part 2

Using above's code will take eternity but instead use dictionary to hold lifespan because there could be same number repeated number of times.

In [38]:
from collections import Counter

Counter(data1)

Counter({1: 197, 3: 26, 4: 30, 2: 28, 5: 19})

1 is repeated 197 times so looping through each of 1s is a bad idea. 

In [39]:
lifes = dict(Counter(data1))

days = 256
for day in range(1, days+1):
    lifes = {l: (0 if lifes.get(l+1) is None else lifes.get(l+1)) for l in range(-1, 8)}
    # make all 8s -1
    lifes[8] = lifes[-1]
    # add new lifes that are exhausted
    lifes[6] += lifes[-1]
    # reset exhausted lifes
    lifes[-1] = 0 
    
print(sum(lifes.values()))

1740449478328


In [40]:
lifes

{-1: 0,
 0: 146206613618,
 1: 179702693822,
 2: 186462311535,
 3: 199000604384,
 4: 237138294070,
 5: 224003964731,
 6: 290215232568,
 7: 119644581976,
 8: 158075181624}

## Day 7

[Here is the link.](https://adventofcode.com/2021/day/7)




In [41]:
data,data1 = get_data(day=7)

In [42]:
data = [int(d) for d in data[0].split(",")]
data1 = [int(d) for d in data1[0].split(",")]

data

[16, 1, 2, 0, 4, 2, 7, 1, 2, 14]

### Part 1

In [43]:
l = len(data1)
f = []

for v in range(l):
    f.append((sum([abs(d-v) for d in data1])))
min(f)        

355989

#### One Liner

In [44]:
min([sum([abs(d-v) for d in data1]) for v in range(len(data1))])

355989

### Part 2

In [45]:
l = len(data1)
f = []

for v in range(l):
    diff = [abs(d-v) for d in data1]
    diffs = sum([sum(list(range(dif+1))) for dif in diff])
    f.append(diffs)
min(f)        


102245489

#### One Liner

In [46]:
min([sum([sum(list(range(abs(d-v)+1))) for d in data1]) for v in range(len(data1))])

102245489

## Day 8

[Here is the link.](https://adventofcode.com/2021/day/8)




In [45]:
import numpy as np
from collections import Counter

data,data1 = get_data(day=8)

In [46]:
data

['be cfbegad cbdgef fgaecd cgeb fdcge agebfd fecdb fabcd edb | fdgacbe cefdb cefbgd gcbe',
 'edbfga begcd cbg gc gcadebf fbgde acbgfd abcde gfcbed gfec | fcgedb cgb dgebacf gc',
 'fgaebd cg bdaec gdafb agbcfd gdcbef bgcad gfac gcb cdgabef | cg cg fdcagb cbg',
 'fbegcd cbd adcefb dageb afcb bc aefdc ecdab fgdeca fcdbega | efabcd cedba gadfec cb',
 'aecbfdg fbg gf bafeg dbefa fcge gcbea fcaegb dgceab fcbdga | gecf egdcabf bgf bfgea',
 'fgeab ca afcebg bdacfeg cfaedg gcfdb baec bfadeg bafgc acf | gebdcfa ecba ca fadegcb',
 'dbcfg fgd bdegcaf fgec aegbdf ecdfab fbedc dacgb gdcebf gf | cefg dcbef fcge gbcadfe',
 'bdfegc cbegaf gecbf dfcage bdacg ed bedf ced adcbefg gebcd | ed bcgafe cdgba cbgef',
 'egadfb cdbfeg cegd fecab cgb gbdefca cg fgcdab egfdb bfceg | gbdfcae bgc cg cgb',
 'gcafb gcf dcaebfg ecagb gf abcdeg gaef cafbge fdbac fegbdc | fgae cfgab fg bagce']

In [47]:
signals, output = data[0].split("|")

signals = [s.strip() for s in signals.strip().split(" ")]
output = [s.strip() for s in output.strip().split(" ")]
signals, output

(['be',
  'cfbegad',
  'cbdgef',
  'fgaecd',
  'cgeb',
  'fdcge',
  'agebfd',
  'fecdb',
  'fabcd',
  'edb'],
 ['fdgacbe', 'cefdb', 'cefbgd', 'gcbe'])

In [58]:
cnts = {2:1, 4:4, 3:7, 7:8}
sol1 = 0


for row in data1:
    signals, output = row.split("|")
    signals = [s.strip() for s in signals.strip().split(" ")]
    output = [s.strip() for s in output.strip().split(" ")]
    
    for o in output:
        l = len(o)
        if cnts.get(l) is not None:
            sol1+=1
sol1    

330

In [60]:
def permutation(li: list):
    all_ps = set()
    psl = np.prod(np.linspace(1,len(li), len(li)).astype(int))
    
    
    while len(all_ps)!=psl:
        curr_ps = np.random.choice(range(len(li)), len(li), replace=False)
        curr_ps = "".join([li[i] for i in curr_ps])
        all_ps.add(curr_ps)
    return all_ps

all_ps = permutation("abcdefg")

d = {
    "abcefg": 0,
    "cf": 1,
    "acdeg": 2,
    "acdfg": 3,
    "bcdf": 4,
    "abdfg": 5,
    "abdefg": 6,
    "acf": 7,
    "abcdefg": 8,
    "abcdfg": 9,
}

cnts = {2:1, 4:4, 3:7, 7:8}

sol1 = 0
sol2 = 0

for row in data1:
    signals, output = row.split("|")
    signals = [s.strip() for s in signals.strip().split(" ")]
    output = [s.strip() for s in output.strip().split(" ")]    
    
    for o in output:
        l = len(o)
        if ls.get(l):
            
            sol1+=1
        
    for pr in all_ps:
        to = str.maketrans("abcdefg", pr)
        ts = ["".join(sorted(sig.translate(to))) for sig in signals]
        top = ["".join(sorted(op.translate(to))) for op in output]

        if all(code in d for code in ts):
            sol2 += int("".join(str(d[code]) for code in top))            
            break
sol1, sol2  

(330, 1010472)

## Day 9

[Here is the link.](https://adventofcode.com/2021/day/9)




In [ ]:
import numpy as np
data,data1 = get_data(day=9)

In [ ]:
dl = len(data[0])
dt = np.array([int(d) for dt in data for d in dt])
dt = dt.reshape(-1, dl)
dt

In [ ]:
dl = len(data1[0])
dt = np.array([int(d) for dt in data1 for d in dt])
dt = dt.reshape(-1, dl)
dt

In [ ]:
nums = []
pos = []
dc = len(dt[0])
dr = len(dt)
for r in range(len(dt)):
    for c in range(len(dt[0])):
        if r==0:
            if c==0:
                if dt[r,c]<dt[r+1, c] and dt[r,c]<dt[r, c+1]:
                    nums.append(dt[r,c])
                    pos.append((r,c))
            elif c==dc-1:
                if dt[r,c]<dt[r+1, c] and dt[r,c]<dt[r, c-1]:
                    nums.append(dt[r,c])
                    pos.append((r,c))
            else:
                if dt[r,c]<dt[r+1, c] and dt[r,c]<dt[r, c+1] and dt[r,c]<dt[r, c-1]:
                    nums.append(dt[r,c])
                    pos.append((r,c))
        elif r==dr-1:
            if c==0:
                if dt[r,c]<dt[r-1, c] and dt[r,c]<dt[r, c+1]:
                    nums.append(dt[r,c])
                    pos.append((r,c))
            elif c==dc-1:
                if dt[r,c]<dt[r-1, c] and dt[r,c]<dt[r, c-1]:
                    nums.append(dt[r,c])
                    pos.append((r,c))
            else:
                if dt[r,c]<dt[r-1, c] and dt[r,c]<dt[r, c+1] and dt[r,c]<dt[r, c-1]:
                    nums.append(dt[r,c])
                    pos.append((r,c))
        else:
            if c==0:
                if dt[r,c]<dt[r-1, c] and dt[r,c]<dt[r, c+1] and dt[r,c]<dt[r+1, c]:
                    nums.append(dt[r,c])
                    pos.append((r,c))
            elif c==dc-1:
                if dt[r,c]<dt[r-1, c] and dt[r,c]<dt[r, c-1] and dt[r,c]<dt[r+1, c]:
                    nums.append(dt[r,c])
                    pos.append((r,c))
            else:
                if dt[r,c]<dt[r-1, c] and dt[r,c]<dt[r, c+1] and dt[r,c]<dt[r, c-1] and dt[r,c]<dt[r+1, c]:
                    nums.append(dt[r,c])
                    pos.append((r,c))
                
                
nums

In [ ]:
sum(nums)+len(nums)

In [ ]:
from scipy import ndimage

label, num_label = ndimage.label(dt < 9)
size = np.bincount(label.ravel())

top3 = sorted(size[1:], reverse=True)[:3]
print(np.prod(top3))

## Day 10

[Here is the link.](https://adventofcode.com/2021/day/10)




In [52]:
data,data1=get_data(day=10)

In [53]:

table = {
    ")": 3,
    "]": 57,
    "}": 1197,
    ">": 25137}

pair = {"(":")","{":"}", "[":"]", "<":">"}


corruptions = []
rem = []
for i,r in enumerate(data):
    stack = []
    is_corr=False
    for c in r:
        if c in pair:
            stack.append(pair[c])
        elif stack.pop() != c:
            print(f"Corrupted {c} at row {i}")
            corruptions.append(c)
            is_corr=True
            break
    if is_corr==False and len(stack)>0:
        rem.append(stack)
            
            
corr = dict(Counter(corruptions))
sum([table[k]*v for k,v in corr.items()])

Corrupted } at row 2
Corrupted ) at row 4
Corrupted ] at row 5
Corrupted ) at row 7
Corrupted > at row 8


26397

In [71]:
mult = {")": 1,
"]": 2,
"}": 3,
">": 4}
all_total=[]
for row in rem:
    s = 0
    for i,c in enumerate(row):
        s+=5**i*mult[c]
    all_total.append(s)
at = sorted(all_total)
at[len(at)//2]

288957

## Day 11

[Here is the link.](https://adventofcode.com/2021/day/11)




In [49]:
import numpy as np
data,data1=get_data(day=11)

In [50]:
data

['5483143223',
 '2745854711',
 '5264556173',
 '6141336146',
 '6357385478',
 '4167524645',
 '2176841721',
 '6882881134',
 '4846848554',
 '5283751526']

In [57]:
darr = np.array([int(d) for dt in data1 for d in dt])
darr = darr.reshape((10, 10))
darr

array([[1, 5, 5, 3, 4, 2, 1, 2, 8, 8],
       [5, 2, 5, 5, 3, 8, 4, 8, 8, 2],
       [1, 2, 2, 4, 3, 1, 5, 7, 3, 2],
       [4, 2, 5, 8, 2, 4, 2, 2, 7, 4],
       [1, 6, 5, 8, 5, 6, 4, 2, 1, 6],
       [6, 8, 7, 2, 6, 5, 1, 1, 8, 2],
       [5, 7, 7, 5, 5, 5, 2, 2, 3, 8],
       [5, 6, 2, 2, 5, 4, 5, 1, 7, 2],
       [8, 7, 6, 6, 6, 7, 2, 3, 1, 8],
       [2, 1, 7, 8, 3, 7, 4, 8, 3, 5]])

In [66]:

adj = [(i,j) for i in range(-1, 2) for j in range(-1,2) if i!=0 or j!=0]
window = {(i,j):darr[i][j] for i in range(10) for j in range(10)}

flashes = 0
i=0
previous = set()

while len(previous)<len(window):
    previous = set()
    window = {k:v+1 for k, v in window.items()}
    while True:
        if sum(v>9 for k,v in window.items() if k not in previous)==0:
            break
            
        for k,v in window.items():
            if k not in previous and v>9:
                previous.add(k)
                for ad in [(k[0]+i,k[1]+j) for i,j in adj if (k[0]+i,k[1]+j) in window]:
                    window[ad]+=1
    
    flashes+=len(previous)
    window.update({k:0 for k in previous})
    i+=1
    if i==100:
        print(f"Part 1: {flashes}")
    
print(f"Part 2: {i}")

Part 1: 1679
Part 2: 519


## Day 12

[Here is the link.](https://adventofcode.com/2021/day/12)

[Took help from here](https://www.reddit.com/r/adventofcode/comments/rehj2r/comment/ho8zlrj/?utm_source=share&utm_medium=web2x&context=3)

In [13]:
data,data1 = get_data(day=12)

In [14]:
class Solver:
    def __init__(self, data):
        self.paths = {}
        self.data = data
        self.visited = set()
        
        self.prepare_paths()
        print(self.solve(part="1"))
        print(self.solve(part="2"))
        
    def prepare_paths(self):
        for d in self.data:
            l,r = d.split("-")
            if self.paths.get(l):
                self.paths[l].append(r)
            else:
                self.paths[l] = [r]
            if self.paths.get(r):
                self.paths[r].append(l)
            else:
                self.paths[r]=[l]
    
    def solve(self, curr_cave="start", part="1"):
        if (curr_cave=="end"):
            return 1
        if curr_cave.islower():
            self.visited.add(curr_cave)
        
        ways_count = sum([self.solve(cave, part) for cave in self.paths[curr_cave] if cave not in self.visited])
        ways_count += 0 if part!="2" else sum([self.solve(cave, cave) for cave in self.paths[curr_cave] if cave in self.visited and cave != "start"])
        
        if (curr_cave != part): self.visited.discard(curr_cave)
        return ways_count
        
                
        
        
s = Solver(data1)

5457
128506


## Day 13

[Here is the link.](https://adventofcode.com/2021/day/13)




In [123]:
import numpy as np
data1, data = get_data(day=13)

In [124]:
data

['745,436',
 '810,236',
 '487,101',
 '938,527',
 '972,747',
 '63,284',
 '979,179',
 '1069,204',
 '170,318',
 '863,166',
 '214,68',
 '370,207',
 '567,544',
 '930,452',
 '1220,396',
 '666,294',
 '78,115',
 '20,353',
 '769,313',
 '992,367',
 '1263,228',
 '965,466',
 '311,851',
 '706,280',
 '1176,309',
 '141,101',
 '972,539',
 '830,638',
 '726,208',
 '388,63',
 '43,614',
 '1153,263',
 '920,528',
 '567,722',
 '693,464',
 '634,211',
 '301,49',
 '152,491',
 '967,603',
 '377,873',
 '1039,648',
 '400,828',
 '1176,217',
 '425,851',
 '730,688',
 '940,511',
 '1208,0',
 '793,71',
 '520,25',
 '1208,3',
 '17,60',
 '1166,791',
 '480,630',
 '867,616',
 '194,113',
 '574,1',
 '1054,880',
 '296,297',
 '763,353',
 '945,16',
 '284,831',
 '487,886',
 '808,480',
 '1149,303',
 '689,547',
 '591,162',
 '649,711',
 '755,386',
 '986,502',
 '502,491',
 '937,287',
 '692,625',
 '1208,782',
 '1302,591',
 '1014,821',
 '949,32',
 '487,738',
 '1044,528',
 '1076,739',
 '75,296',
 '768,68',
 '621,155',
 '498,682',
 '472,75

In [125]:
dots = [list(map(int, f.split(","))) for f in data[:data.index("")]]
folds = data[data.index("")+1:]
folds = [f.split("along ")[1].split("=") for f in folds]
folds = [(f[0], int(f[1])) for f in folds]
dots

[[745, 436],
 [810, 236],
 [487, 101],
 [938, 527],
 [972, 747],
 [63, 284],
 [979, 179],
 [1069, 204],
 [170, 318],
 [863, 166],
 [214, 68],
 [370, 207],
 [567, 544],
 [930, 452],
 [1220, 396],
 [666, 294],
 [78, 115],
 [20, 353],
 [769, 313],
 [992, 367],
 [1263, 228],
 [965, 466],
 [311, 851],
 [706, 280],
 [1176, 309],
 [141, 101],
 [972, 539],
 [830, 638],
 [726, 208],
 [388, 63],
 [43, 614],
 [1153, 263],
 [920, 528],
 [567, 722],
 [693, 464],
 [634, 211],
 [301, 49],
 [152, 491],
 [967, 603],
 [377, 873],
 [1039, 648],
 [400, 828],
 [1176, 217],
 [425, 851],
 [730, 688],
 [940, 511],
 [1208, 0],
 [793, 71],
 [520, 25],
 [1208, 3],
 [17, 60],
 [1166, 791],
 [480, 630],
 [867, 616],
 [194, 113],
 [574, 1],
 [1054, 880],
 [296, 297],
 [763, 353],
 [945, 16],
 [284, 831],
 [487, 886],
 [808, 480],
 [1149, 303],
 [689, 547],
 [591, 162],
 [649, 711],
 [755, 386],
 [986, 502],
 [502, 491],
 [937, 287],
 [692, 625],
 [1208, 782],
 [1302, 591],
 [1014, 821],
 [949, 32],
 [487, 738],
 [1

In [126]:
dots = np.array(dots)
dots

array([[745, 436],
       [810, 236],
       [487, 101],
       ...,
       [ 85, 690],
       [634, 894],
       [618, 625]])

In [127]:
window = np.zeros(dots.max(axis=0)+3)
window

array([[0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.]])

In [128]:
# window[:]="."
for c in dots:
    window[c[0], c[1]] = 1
window = window.T
window

array([[0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.]])

In [141]:
tw = window.copy()
# print(tw)
for f in folds:
    print(f)
    axis,value=f
    cr,cc = tw.shape
    print(tw)
    if axis=="y":
        #fold y axis
        chunk = tw[value+1:-2]
        # print(value-crs,chunk.shape, chunk)
        crs,ccs = chunk.shape
        tw[np.abs(value-crs):value] += chunk[::-1]
        tw = tw[:value]
        tw = np.append(tw, np.zeros((2, tw.shape[1])), axis=0)
        #break
        
    else:
        # fold x axis
        chunk = tw[:, value+1:-2]
        crs,ccs = chunk.shape
        print(value-crs,chunk.shape, chunk)
        tw[:, abs(value-ccs):value] += chunk[:,::-1]
        tw = tw[:, :value]
        tw = np.append(tw, np.zeros((tw.shape[0], 2)), axis=1)
    print(f"Dots: {np.sum(tw>0)}")
tw

('x', 655)
[[0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 ...
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]]
-242 (897, 655) [[0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 ...
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]]
Dots: 720
('y', 447)
[[0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 ...
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]]
Dots: 594
('x', 327)
[[0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 ...
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]]
-122 (449, 327) [[0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 ...
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]]
Dots: 510
('y', 223)
[[0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 ...
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0

array([[ 0.,  2., 11.,  0.,  0., 10.,  0.,  0.,  6.,  0., 16.,  8.,  8.,
         0.,  0.,  3.,  2.,  4.,  0.,  0., 23.,  6.,  1.,  0.,  0.,  0.,
         4., 13.,  0.,  0., 10.,  0.,  0.,  3.,  0.,  2.,  2.,  9., 17.,
         0.,  0.,  0.],
       [ 2.,  0.,  0., 16.,  0.,  9.,  0.,  0.,  5.,  0., 10.,  0.,  0.,
         2.,  0., 13.,  0.,  0., 16.,  0., 10.,  0.,  0.,  9.,  0.,  3.,
         0.,  0., 23.,  0.,  5.,  0.,  0., 19.,  0.,  0.,  0.,  0.,  7.,
         0.,  0.,  0.],
       [ 2.,  0.,  0., 10.,  0., 23., 12.,  3.,  5.,  0.,  8.,  0.,  0.,
         5.,  0.,  6.,  0.,  0.,  6.,  0.,  8.,  0.,  0., 12.,  0.,  2.,
         0.,  0.,  3.,  0.,  2.,  0.,  0., 10.,  0.,  0.,  0.,  8.,  0.,
         0.,  0.,  0.],
       [11.,  5.,  3., 17.,  0.,  3.,  0.,  0., 10.,  0.,  5., 13.,  2.,
         0.,  0.,  3.,  8.,  5.,  0.,  0.,  9., 11.,  9.,  0.,  0.,  8.,
        11.,  9.,  7.,  0.,  6.,  0.,  0.,  5.,  0.,  0., 18.,  0.,  0.,
         0.,  0.,  0.],
       [ 4.,  0.,  0., 11., 

In [148]:
print(np.array2string(tw>0, separator='',
    formatter = {'bool':lambda x: ' █'[x]}))

[[ ██  █  █ ███  ███  ███   ██  █  █ ████   ]
 [█  █ █  █ █  █ █  █ █  █ █  █ █  █    █   ]
 [█  █ ████ █  █ █  █ █  █ █  █ █  █   █    ]
 [████ █  █ ███  ███  ███  ████ █  █  █     ]
 [█  █ █  █ █    █ █  █    █  █ █  █ █      ]
 [█  █ █  █ █    █  █ █    █  █  ██  ████   ]
 [                                          ]
 [                                          ]]


<ipython-input-148-6686f362f065>:2: DeprecationWarning: In future, it will be an error for 'np.bool_' scalars to be interpreted as an index
  formatter = {'bool':lambda x: ' █'[x]}))


## Day 14

[Here is the link.](https://adventofcode.com/2021/day/14)




In [93]:
data,data1 = get_data(day=14)

In [98]:
wdata=data1.copy()
polymer = wdata[0]
rule = {v[0]:v[1] for v in [d.split(" -> ") for d in wdata[2:]]}
# rule,wdata

In [95]:
curr_polymer = polymer
i=0
while i< 10:
    tpoly = curr_polymer
#     print(i, tpoly)
    ind = 0
    added = 0
    for k, c in enumerate(tpoly):
        k+=1
        ch = curr_polymer[k-1:k+1]
        
        mc = rule.get(ch)
        if mc:
            tpoly = [c for c in tpoly]
            tpoly.insert(k+added, mc)
            tpoly = "".join(tpoly)
            added+=1
    curr_polymer = tpoly
    i+=1

In [96]:
from collections import Counter
res = dict(Counter(curr_polymer))
res = sorted(res.items(), key=lambda x: x[1], reverse=True)
res[0][1]-res[-1][1]

3284

In [97]:
tmp_poly = Counter(a+b for a,b in zip(polymer, polymer[1:]))
print(tmp_poly)
chars = Counter(polymer)

for _ in range(40):
    tmp = Counter()
    for (c1,c2),value in tmp_poly.items():
        mc = rule[c1+c2]
        tmp[c1+mc] += value
        tmp[mc+c2] += value
        chars[mc] += value
    tmp_poly=tmp
max(chars.values()) - min(chars.values())

Counter({'OF': 1, 'FS': 1, 'SV': 1, 'VV': 1, 'VS': 1, 'SF': 1, 'FO': 1, 'OC': 1, 'CB': 1, 'BN': 1, 'NO': 1, 'ON': 1, 'NH': 1, 'HK': 1, 'KF': 1, 'FH': 1, 'HN': 1, 'NP': 1, 'PK': 1})


4302675529689

## Day 15

[Here is the link.](https://adventofcode.com/2021/day/15)



In [51]:
import numpy as np

data,data1 = get_data(15)

In [58]:
data = np.array([int(i) for dt in data for i in dt ]).reshape(-1, len(data[0]))
data
data1 = np.array([int(i) for dt in data1 for i in dt ]).reshape(-1, len(data1[0]))

In [75]:
from skimage import graph
import numpy as np
window = data1.copy()

rs,cs = window.shape

cost = graph.MCP(window, fully_connected=False)
cost.find_costs(starts = [(0,0)])

journey = [window[pos] for pos in  cost.traceback((rs-1,cs-1))[1:]]
print(f"Part1: {sum(journey)}")

# 5times bigger
new_window = window.copy()
nrow = np.hstack([new_window, new_window+1, new_window+2, new_window+3, new_window+4])
new_window = np.vstack([nrow,nrow+1,nrow+2,nrow+3,nrow+4])
rs,cs = new_window.shape

new_window%=9
new_window[new_window==0]=9

cost = graph.MCP(new_window, fully_connected=False)
cost.find_costs(starts = [(0,0)])

journey = [new_window[pos] for pos in  cost.traceback((rs-1,cs-1))[1:]]
print(f"Part2: {sum(journey)}")

Part1: 581
Part2: 2916



## Day 16

[Here is the link.](https://adventofcode.com/2021/day/16)




## Day 17

[Here is the link.](https://adventofcode.com/2021/day/17)




## Day 18

[Here is the link.](https://adventofcode.com/2021/day/18)




## Day 19

[Here is the link.](https://adventofcode.com/2021/day/19)




## Day 20

[Here is the link.](https://adventofcode.com/2021/day/20)




## Day 21

[Here is the link.](https://adventofcode.com/2021/day/21)




## Day 22

[Here is the link.](https://adventofcode.com/2021/day/22)




## Day 23

[Here is the link.](https://adventofcode.com/2021/day/23)




## Day 24

[Here is the link.](https://adventofcode.com/2021/day/24)




## Day 25

[Here is the link.](https://adventofcode.com/2021/day/25)


# Preparation Section

### Push Code

In [149]:
! git add .
! git commit -m "solved day13"
! git push origin master

The file will have its original line endings in your working directory
The file will have its original line endings in your working directory


[master 3795403] solved day13
 2 files changed, 3086 insertions(+), 22 deletions(-)
 rewrite 2021/data/day13.txt (100%)


To https://github.com/q-viper/Adevent-Of-Code.git
   ae31c75..3795403  master -> master


## Prepare data files for all days

In [ ]:
for i in range(2, 26):
    with open(f"data/day{i}.txt","w") as fp:
        fp.writelines("Split From Here")
    

In [ ]:

for i in range(4, 26):
    p = f"""
    ## Day {i}
    [Here is the problem link.](https://adventofcode.com/2021/day/{i})
    
    ### Part 1
    ```python
    
    ```
    
    ### Part 2
    ```python
    
    ```
    """
    print(p)

In [ ]:
np.random.normal(50,5,100)